toolchain/lex/lex.cpp

```diff
@@ -1104,40 +1103,43 @@ auto Lexer::LexKeywordOrIdentifier(llvm::StringRef source_text,
        .ident_id = buffer_.value_stores_->identifiers().Add(identifier_text)});
 }
 
-auto Lexer::LexKeywordOrIdentifierMaybeRaw(llvm::StringRef source_text,
-                                           ssize_t& position) -> LexResult {
-  CARBON_CHECK(source_text[position] == 'r');
-  // Raw identifiers must look like `r#<valid identifier>`, otherwise it's an
-  // identifier starting with the 'r'.
-  // TODO: Need to add support for Unicode lexing.
-  if (LLVM_LIKELY(position + 2 >= static_cast<ssize_t>(source_text.size()) ||
-                  source_text[position + 1] != '#' ||
-                  !IsIdStartByteTable[static_cast<unsigned char>(
-                      source_text[position + 2])])) {
-    // TODO: Should this print a different error when there is `r#`, but it
-    // isn't followed by identifier text? Or is it right to put it back so
-    // that the `#` could be parsed as part of a raw string literal?
-    return LexKeywordOrIdentifier(source_text, position);
+auto Lexer::LexHash(llvm::StringRef source_text, ssize_t& position)
+    -> LexResult {
+  // For `r#`, we already lexed an `r` identifier token. Detect that case and
+  // replace that token with a raw identifier. We do this to keep identifier
+  // lexing as fast as possible.
+
+  // Look for the `r` token. Note that this is always in bounds because we
+  // create a start of file token.
+  TokenIndex maybe_r_index = buffer_.tokens().end()[-1];
```

Comment:
As mentioned, this feels oddly expensive versus `token_infos_.back()` -- creates an iterator range, offsets by one, then uses the result. Not sure how much of that would be optimized away.